lib/rules/v-if-else-key.js

```diff
@@ -0,0 +1,378 @@
+/**
+ * @author Felipe Melendez
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+// =============================================================================
+// Requirements
+// =============================================================================
+
+const utils = require('../utils')
+const casing = require('../utils/casing')
+const reservedNamesInVue2 = require('../utils/vue2-builtin-components')
+
+// =============================================================================
+// Constants and Variables
+// =============================================================================
+
+/**
+ * Curated list of standard HTML elements that serve as containers.
+ * We're not using an existing library util because we are only concerned about HTML elements that
+ * have the capacity to wrap other components.
+ *
+ * @type {Array<string>}
+ */
+const htmlContainerElements = [
```

Comment:
Please remove this list (and everything that belongs to it); it should be unnecessary now with the `isCustomComponent` check.

Reply:
hmm, I don't think we can do that, the rule maintains a component usage stack (componentUsageStack), which is a stack of maps where each map represents a scope level, tracking the usage of custom Vue components at that level. When the rule encounters a container element (determined by isContainerElement), it pushes a new map onto the componentUsageStack, essentially entering a new scope level. This is done because container elements can have their own scope of children, and the rule needs to track component usage separately for each scope.

Reply:
This level of scope tracking enables us to discern that the following structure is acceptable:
```
<template>
    <div>
        <div v-if="some-condition" >
            <CustomComponent />
        </div>
        <CustomComponent v-else />
    </div>
</template>
```

And it allows us to identify that this structure, conversely, is problematic:
```
<template>
    <div>
        <CustomComponent v-if="someCondition" />
        <CustomComponent v-else-if="otherCondition" />
        <CustomComponent v-else />
    </div>
</template>
```

Reply:
But again, it doesn't matter if the tag in your first example is a `<div>` or a `<span>`. The only tag that would need to be handled specially would be `<template>`.

Reply:
Sorry Flo, are we on the same page? I'm not sure what you mean (why are you singling out only template, component, and slot?). Perhaps I should not have used divs to illustrate my point?

I'm talking about differentiating between scope levels to ensure that we only add keys to conditionally rendered repeated components that are at the same scope level. 

Here is a more generic illustration:
```
<some-container>
    <ScopeOne v-if="conditionOne"> 
        <ScopeTwo v-if="conditionTwo" /> <--- this should have a key
        <ScopeTwo v-else /> <--- this should have a key
    </ScopeOne>
    <ScopeTwo v-else /> <--- this should not have a key
</some-container>
```

Reply:
I agree, but what do you need the `htmlContainerElements` variable for then?

Reply:
ah, thank you for being patient, I was relying on memory and variable names to construct my mental model of the rules behavior, and both threw me off (several months have passed since I built this thing). Now I have done what I should have done from the beginning and studied my own code, and you are absolutely right, I have refactored. Once again thank you for your attention to detail. 
I have refactored and added a test to specifically ensure that what we have discussed here is covered by tests:
[c659339](https://github.com/vuejs/eslint-plugin-vue/pull/2280/commits/c659339aca07875daedea6014c14d24d122be087)

[c89efd8](https://github.com/vuejs/eslint-plugin-vue/pull/2280/commits/c89efd80cc10ff3d67a2570155141ea283bfb83b)

Reply:
by the way, I had already captured that idea in a previous test, but this new one makes the concept more quickly evident, I can remove it if you think it's unnecessary

Reply:
Thanks for checking it again and adding a test!
This definitely looks easier now :slightly_smiling_face: